Hw09

tikalet opened the pull request.

tikalet committed:
add project

tikalet committed:
fix service tests

tikalet committed:
add comment pages

tikalet committed:
add tests for book controller

tikalet committed:
add tests for comment controller

vlad-iv commented on hw09/src/main/java/ru/otus/hw/controller/BookController.java:
```diff
@@ -0,0 +1,78 @@
+        return "book_edit";
+    }
+
+    @PostMapping(value = "/book_edit")
```
Можно указать путь /book/{id}

vlad-iv commented on hw09/src/main/java/ru/otus/hw/controller/BookController.java:
```diff
@@ -0,0 +1,78 @@
+        return "books";
+    }
+
+    @GetMapping(value = "/book", params = "id")
```
Можно указать путь /book/{id}

vlad-iv commented on hw09/src/main/java/ru/otus/hw/controller/BookController.java:
```diff
@@ -0,0 +1,78 @@
+        return "redirect:/";
+    }
+
+    @PostMapping(value = "/book_delete", params = "id")
```
Можно указать путь /book/{id}/delete

vlad-iv commented on hw09/src/main/java/ru/otus/hw/controller/BookController.java:
```diff
@@ -0,0 +1,78 @@
+        return "book_new";
+    }
+
+    @PostMapping(value = "/book_save")
```
Можно указать путь /book

vlad-iv commented on hw09/src/main/java/ru/otus/hw/dto/AuthorDto.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public String toString() {
```
В Dto можно удалить toString , equals и тд.
а вместо всех аннотаций указать Data

vlad-iv commented on hw09/src/main/java/ru/otus/hw/dto/BookDto.java:
```diff
@@ -0,0 +1,62 @@
+
+    private GenreDto genre;
+
+    public static BookDto fromModel(Book book) {
```
(опционально) можно создать компоненты мапперы по преобразования сущностей и dto
BookMapper

vlad-iv commented on hw09/src/main/java/ru/otus/hw/exceptions/EntityNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.otus.hw.exceptions;
+
+public class EntityNotFoundException extends RuntimeException {
```
Можно переименовать в NotFoundException

vlad-iv commented on hw09/src/main/java/ru/otus/hw/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ModelAndView handeNotFoundException(EntityNotFoundException ex) {
```
Осталось указать что код ответа 404

И добавить обработчик для Exception c кодом ответа 500

vlad-iv commented on hw09/src/main/java/ru/otus/hw/services/BookService.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<BookDto> findAll();
+
+    BookDto insert(String title, long authorId, long genreId);
```
Вместо String title, long authorId, long genreId
Можно применить новое Dto
BookCreateDto
аналогично для update применить BookUpdateDto

(опционально) указать аннотации валидации над полями в Dto + Valid перед ним в контроллере

vlad-iv replied:
все insert методы можно переименовать в create

tikalet committed:
fix book pages

- add mappers
- rename EntityNotFoundException  to NotFoundException
- change book controller's method
- add error code
- fix BookControllerTest
- delete book new page

tikalet committed:
fix comment controller

tikalet committed:
add saveDto

tikalet committed:
fix save book

- add favicon
- change error code
- remove 'static' for mapper method

tikalet committed:
fix save comment

tikalet committed:
add validation

tikalet committed:
add tests for validation

vlad-iv commented on hw09/src/main/java/ru/otus/hw/controller/BookController.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @PostMapping(value = "/book")
+    public String saveBook(@Valid @ModelAttribute("book") BookSaveDto book, BindingResult bindingResult, Model model) {
```
Создание и обновление могут требовать разную валидацию
Предлагаю в следующей работе разбить на 2 действия

vlad-iv commented on hw09/src/main/java/ru/otus/hw/dto/BookSaveDto.java:
```diff
@@ -0,0 +1,23 @@
+    @Size(min = 0, max = 254, message = "Max symbols are 254")
+    private String title;
+
+    private long authorId;
```
Во входящем Dto ожидается что будут классы обертки
Что бы можно было проверить наличие поля
Так же можно добавить что поля автор / жанра обязательные или же выбрать, что опциональные

ид при создании должно отсутствовать, а при обновлении обязательное
